### PR TITLE
Url to iframe tests

### PR DESCRIPTION
**Pull request by jrhender**
This PR:
- Adds `jest` dependencies to use `jest` for tests. Jest selected because it provides full functionality for units tests (e.g. test runner, mocking, assertions) in a single package
- Adds npm script to execute tests
- Adds `npm run test` to `pr-check.yml`
- Adds note to README on running tests
- Adds simple tests for `urlToIframe` function

### CODE REVIEW

File: .github/workflows/unit-test.yml
Comment (Aprillion) [inline .github/workflows/unit-test.yml:15]: please move the `npm run test` into `pr-check.yml` file, not a new GH action

File: .gitignore
Comment (Aprillion) [inline .gitignore:14]: unless you store any secrets in this file, I would be fine including it in git for new developers to have better VSCode integration

File: app/server-utils/url-to-iframe.spec.ts
Comment (Aprillion) [inline app/server-utils/url-to-iframe.spec.ts:14]: did you run prettier for the new files (`npm run lint`)? I believe it should have added newline at the end of files automatically...

otherwise I like the first test 👍

**Review by Aprillion — APPROVED** [review]

File: .github/workflows/unit-test.yml
Comment (jrhender) [reply .github/workflows/unit-test.yml]: Yes, you're right, I think that is simpler for now.
Done here: https://github.com/StampyAI/stampy-ui/pull/193/commits/fed9bc3f2749d6f28bdf354ff0560fe003193ddd

File: .gitignore
Comment (jrhender) [reply .gitignore]: Ok, sounds good. Added here: https://github.com/StampyAI/stampy-ui/pull/193/commits/0307b767e45c247aed533982e6d45fa958df9661

File: app/server-utils/url-to-iframe.spec.ts
Comment (jrhender) [reply app/server-utils/url-to-iframe.spec.ts]: Thanks, yes, you were right, it needed a prettier run https://github.com/StampyAI/stampy-ui/pull/193/commits/bbb7a85290548f2b6f459252704e1b55ae1efa26

File: README.md
Comment (Aprillion) [inline README.md:72]: 👍

**Review by Aprillion — APPROVED** [review]

File: .vscode/launch.json
Comment (mruwnik) [inline .vscode/launch.json:1]: should this be here? It looks like an editor specific file. Could you also add it to .gitignore?

File: app/server-utils/url-to-iframe.spec.ts
Comment (mruwnik) [inline app/server-utils/url-to-iframe.spec.ts:12]: the trailing `</a>` should probably be removed

File: app/server-utils/url-to-iframe.spec.ts
Comment (mruwnik) [inline app/server-utils/url-to-iframe.spec.ts:18]: could you also test these?:
* `[https://aisafety.world/?bla=bla](https://aisafety.world/?bla=bla)`
* `[https://aisafety.world/](https://aisafety.world/?bla=bla)`
* `[https://aisafety.world/?bla=bla](https://aisafety.world/)`
* `[http://aisafety.world/](http://aisafety.world/)`
* `[aisafety.world](http://aisafety.world/)`
* `[aisafety.world/](aisafety.world/)`

File: README.md
Comment (mruwnik) [reply README.md:72]: 👍

**Review by mruwnik — CHANGES_REQUESTED** [review]

**Comment by mruwnik** [comment]
This is really useful - good work!

File: .vscode/launch.json
Comment (mruwnik) [reply .vscode/launch.json:1]: ah. I see @Aprillion wants this - disregard this comment :D

File: app/server-utils/url-to-iframe.spec.ts
Comment (jrhender) [reply app/server-utils/url-to-iframe.spec.ts:18]: Yes, certainly. Thanks for thinking of all these cases.

What should the expected results be for each of them in your view? Should all of them convert into the iframe or just some of them?

I went back and looked at the GH issue for this iframe (https://github.com/StampyAI/stampy-ui/issues/140) and @plexish was saying that "If a full plain URL from the whitelisted domains is in the markdown, it should be transformed into a large (~screen height, full answer width) iframe of that URL."
Re-reading that, I wonder if my initial understanding and implementation isn't quite correct. The current implementation is considering the `href`/`()` portion of the link but maybe this is not relevant. Maybe it just matters that the text/`[]` portion is a whitelisted URL.

With this new understanding, I think the expected behaviour should be:
1. `[https://aisafety.world/?bla=bla](https://aisafety.world/?bla=bla)` -> should render iframe with href of `https://aisafety.world/?bla=bla`
1. `[https://aisafety.world/](https://aisafety.world/?bla=bla)` -> should render iframe with href of `https://aisafety.world/`
1. `[https://aisafety.world/?bla=bla](https://aisafety.world/)`  render iframe with href of `https://aisafety.world/?bla=bla`
1. `[http://aisafety.world/](http://aisafety.world/)` -> should render iframe with href of `https://aisafety.world/`
1. `[aisafety.world](http://aisafety.world/)` -> should **not** render iframe because it's missing URI scheme
1. `[aisafety.world/](aisafety.world/)`  -> should **not** render iframe because it's missing URI scheme

Does the above seem right?

Comment (mruwnik) [reply app/server-utils/url-to-iframe.spec.ts:18]: I'd personally go with both having to be identical. But what you propose is also totally reasonable. The main thing is to have a consistent view

Comment (plexish) [reply app/server-utils/url-to-iframe.spec.ts:18]: I'm okay with either of these schemes a long as it's documented somewhere the editors can read like the formatting question.

Comment (jrhender) [reply app/server-utils/url-to-iframe.spec.ts]: Ok, I think "both having to be indentical" is the easier route for now and I ideally don't want to change the implementation in an "adding tests" PR.

That said, the first test case you suggested doesn't create an iframe with the current implementation. I think the query string breaks it.
> `[https://aisafety.world/?bla=bla](https://aisafety.world/?bla=bla)` -> should render iframe with href of `https://aisafety.world/?bla=bla`

I made it a failing test case: https://github.com/StampyAI/stampy-ui/pull/193/commits/97390534ac31e9f464066b639997ae36285b96e0
We can come back and add support for query strings in a separate PR, IMO

Comment (mruwnik) [reply app/server-utils/url-to-iframe.spec.ts]: nice! @Aprillion would you prefer this to be fixed now, or do you think that urls with query strings should work as they do now? 
@jrhender if you want to fix it, the regex should be `const matches = text.matchAll(/\[(https?:\/\/[\.\w]+?\/?[^\]]*?)\]\(\1\)/gi)`, as it is in the comment

Comment (Aprillion) [reply app/server-utils/url-to-iframe.spec.ts]: I would prefer to fix it, that's the purpose of tests, no strong opinion between following options:
* a) this PR with [it.failing(...)](https://jestjs.io/docs/api#testfailingname-fn-timeout), and the PR author can file a GH issue with the bug and then someone can fix code in next PR, or
* b) 2 separate commits in this PR so that tests are passing before we merge, or
* c) 1 commit with a passing test and fixed code at the same time

just no merging of red tests, please

up to @jrhender what option you feel most comfortable with, though TBH, option b) sounds most straight forward (if the code from above comment fixes the problem without making the other tests unhappy)

Comment (jrhender) [reply app/server-utils/url-to-iframe.spec.ts]: Ah I didn't now about `it.failing`, that's nifty.

Ok, I moved querystring case to be expected to pass and adjusted regex as per mruwnik's tips https://github.com/StampyAI/stampy-ui/pull/193/commits/566a78fc4f18e2ffde2246e583222f8022f7298c

Comment (jrhender) [reply app/server-utils/url-to-iframe.spec.ts]: > as it's documented somewhere the editors can read like the formatting question.

@plexish I'm glad to document this. Do you know where a good place to do this might be? Is there a Q&A writing guide perhaps?

**Review by mruwnik — DISMISSED** [review]

**Comment by jrhender** [comment]
Ah, I see I need to run prettier

File: app/server-utils/url-to-iframe.spec.ts
Comment (plexish) [reply app/server-utils/url-to-iframe.spec.ts]: yep write it in a comment there and I'll include it @jrhender

**Pull request merged**
